lightning/src/ln/outbound_payment.rs

```diff
@@ -301,11 +429,124 @@ impl OutboundPayments {
 			None => PaymentPreimage(entropy_source.get_secure_random_bytes()),
 		};
 		let payment_hash = PaymentHash(Sha256::hash(&preimage.0).into_inner());
-		let onion_session_privs = self.add_new_pending_payment(payment_hash, None, payment_id, &route, entropy_source, best_block_height)?;
+		let onion_session_privs = self.add_new_pending_payment(payment_hash, None, payment_id, &route, Retry::Attempts(0), None, entropy_source, best_block_height)?;
 
-		match self.send_payment_internal(route, payment_hash, &None, Some(preimage), payment_id, None, onion_session_privs, node_signer, best_block_height, send_payment_along_path) {
+		match self.pay_route_internal(route, payment_hash, &None, Some(preimage), payment_id, None, onion_session_privs, node_signer, best_block_height, &send_payment_along_path) {
 			Ok(()) => Ok(payment_hash),
-			Err(e) => Err(e)
+			Err(e) => {
+				self.remove_outbound_if_all_failed(payment_id, &e);
+				Err(e)
+			}
+		}
+	}
+
+	pub(super) fn check_retry_payments<R: Deref, ES: Deref, NS: Deref, SP, IH, FH, L: Deref>(
+		&self, router: &R, first_hops: FH, inflight_htlcs: IH, entropy_source: &ES, node_signer: &NS,
+		best_block_height: u32, logger: &L, send_payment_along_path: SP,
+	)
+	where
+		R::Target: Router,
+		ES::Target: EntropySource,
+		NS::Target: NodeSigner,
+		SP: Fn(&Vec<RouteHop>, &Option<PaymentParameters>, &PaymentHash, &Option<PaymentSecret>, u64,
+		   u32, PaymentId, &Option<PaymentPreimage>, [u8; 32]) -> Result<(), APIError>,
+		IH: Fn() -> InFlightHtlcs,
+		FH: Fn() -> Vec<ChannelDetails>,
+		L::Target: Logger,
+	{
+		loop {
+			let mut outbounds = self.pending_outbound_payments.lock().unwrap();
+			let mut retry_id_route_params = None;
+			for (pmt_id, pmt) in outbounds.iter_mut() {
+				if pmt.is_retryable_now() {
+					if let PendingOutboundPayment::Retryable { pending_amt_msat, total_msat, route_params: Some(params), .. } = pmt {
+						if pending_amt_msat < total_msat {
```

Comment:
Why does this indicate that the payment should be retried? Can't the condition be true when there are only in-flight HTLCs that haven't failed?

Reply:
Chatted offline. `pending_amt_msat` starts at `total_msat`, so should be good.  I think from a readability perspective, moving the initialization of `PendingOutboundPayment::Retryable` into a method on `OutboundPayments` would help.  Essentially the match statement in `add_new_pending_payment`.

Reply:
Oh, `add_new_pending_payment` is already a method on `OutboundPayments` lol... Feel free to disregard.